src/lib/mergeThemes.ts

```diff
@@ -0,0 +1,124 @@
+ * They are flat objects and do not depend on render-time values, such as props.
+ */
+const mergeSiteVariables = (
+  target: ISiteVariables,
+  ...sources: ISiteVariables[]
+): ISiteVariables => {
+  return sources.reduce((acc, next) => ({ ...acc, ...next }), target)
+}
+
+/**
+ * Component variables can be objects, functions, or an array of these.
+ * The functions must be called with the final result of siteVariables, otherwise
+ *   the component variable objects would have no ability to apply siteVariables.
+ * Therefore, componentVariables must be resolved by the component at render time.
+ * We instead pass down call stack of component variable functions to be resolved later.
+ */
+const mergeComponentVariables = (
+  target: IThemeComponentVariablesInput,
+  ...sources: IThemeComponentVariablesInput[]
+): IThemeComponentVariablesPrepared => {
+  const displayNames = _.union(_.keys(target), ..._.map(sources, _.keys))
+
+  return sources.reduce((acc, next) => {
+    return displayNames.reduce((componentVariables, displayName) => {
+      // Break references to avoid an infinite loop.
+      // We are replacing functions with new ones that calls the originals.
+      const originalTarget = acc[displayName]
+      const originalSource = next[displayName]
+
+      componentVariables[displayName] = siteVariables => {
+        return {
+          ...callable(originalTarget)(siteVariables),
+          ...callable(originalSource)(siteVariables),
+        }
+      }
+
+      return componentVariables
+    }, {})
+  }, target)
+}
+
+/**
+ * See mergeComponentVariables() description.
+ * Component styles adhere to the same pattern as component variables, except
+ *   that they return style objects.
+ */
+const mergeComponentStyles = (
+  target: IThemeComponentStylesInput,
+  ...sources: IThemeComponentStylesInput[]
+): IThemeComponentStylesPrepared => {
+  const initial: IThemeComponentStylesPrepared = _.mapValues(target, stylesByPart => {
+    return _.mapValues(stylesByPart, callable)
+  })
+
+  return sources.reduce<IThemeComponentStylesPrepared>((acc, next) => {
+    _.forEach(next, (stylesByPart, displayName) => {
+      acc[displayName] = acc[displayName] || {}
+
+      _.forEach(stylesByPart, (partStyle, partName) => {
+        // Break references to avoid an infinite loop.
+        // We are replacing functions with a new ones that calls the originals.
+        const originalTarget = acc[displayName][partName]
+        const originalSource = next[displayName][partName]
+
+        acc[displayName][partName] = styleParam => {
+          return _.merge(callable(originalTarget)(styleParam), callable(originalSource)(styleParam))
+        }
+      })
+    })
+
+    return acc
+  }, initial)
+}
+
+const mergeRTL = (target, ...sources) => {
+  return !!sources.reduce((acc, next) => {
+    return typeof next === 'boolean' ? next : acc
+  }, target)
+}
+
+const mergeThemes = (...themes: IThemeInput[]): IThemePrepared => {
```

Comment:
Nit: I think this would be more simple if it took an array of themes.

1) It doesn't really take multiple arguments. It takes a single _collection_ of themes. Representing this with an appropriate data type makes this obvious.

2) It becomes possible to add additional arguments if needed.

3) It makes it easier to provide a list of themes to this function. You can do `mergeThemes(...myListOfThemes)`, but I would ask "why?" when you can just do `mergeThemes(myListOfThemes)`.

4) Least importantly, it saves you casting `arguments` to an array.

Justified myself because you asked :).